### DISCUSSION TITLE
Decaying an array of activies

### DISCUSSION THREAD

**Discussion by buresjan** [radioactivedecay/radioactivedecay#95] [Ideas]
Hello!

Big thanks for creating and actively managing this great open-source project!

Me and my team love this package and we want to use it in our project for decay calculations. At some point in our project, however, we need to be able to decay an activity field (NumPy ndarray) of a single nuclide. To my knowledge, constructing an `rd.Inventory` containing list / array of activities as values for nuclides (the keys) and decaying it is currently not supported.

Using [numpy.vectorize](https://numpy.org/doc/stable/reference/generated/numpy.vectorize.html) I was able to create a simple function that enables decaying entire activity field for an inventory with a single nuclide. It would also be easy to implement it this way for an inventory with multiple nuclides where the arrays would have the same dimension (shape). So one would be able to do something like this:

```python
inv_0 = rd.Inventory({'U-238': ary1, 'U-235': ary2, 'U-234': ary3})
inv_1 = inv_t0.decay(1E6, 'y')
```
where `ary1, ary2, ary3` are e.g. NumPy ndarrays with the same shape. The arrays are then decayed "cell-by-cell", the activity fields of the products also correspond "cellwise".

I mentioned that I already implemented a simplified solution to this problem, but I believe integrating this feature (decaying of arrays) into the `radioactivedecay` package itself could be useful for more users. Therefore, I wanted to ask whether you would consider adding this feature.

Thanks!

**Comment by alexmalins** [comment]
Hey again @buresjan 👋 

I understand the need, but I guess I've reservations about the second suggested solution like:

```
inv_0 = rd.Inventory({'U-238': ary1, 'U-235': ary2, 'U-234': ary3})
inv_1 = inv_t0.decay(1E6, 'y')
```

The issue this causes is it will break the Python API of `Inventory` objects. For instance if the user tried to plot - `inv_0.plot(1e5, 'y')` - what should happen? Should the code try and create one plot with a set of lines for each index in the ndarrays? Should it plot N figures where N is the size of the ndarrays?

So it complicates things, as Inventory objects will have to change to cope with two sets of scenarios: when the quantities of the nuclides in the content are i) scalars & ii) vectors.

A cleaner solution which is I guess is similar to your first `numpy.vectorize()` solution is just to consider an array of N Inventory's like `t_0 = [inv_0, inv_1, inv_2... , inv_n]`.

You can setup such a system and decay it using quite concise Python code using Python's inbuilt map function:

```
nuclides = ['U-238', 'U-235', 'U-234']
acts = np.array([[1., 2., 3.], [4., 5., 6.], [7., 8., 9.]])  # this is effectively [ary1, ary2, ary3] in second example
t_0 = list(map(lambda x: rd.Inventory(dict(zip(nuclides, x))), acts.T))
t_1 = list(map(lambda x: x.decay(1E6, 'y'), t_0))
```

The result `t_1` is a list of inventories, the 0th element is the Inventory {'U-238': 1., 'U-235': 4., 'U-234': 7.} decayed by 1 million years.

You could do similar with list comprehensions if map's are not your thing. Likewise it's easy enough to take the results in t_1 and convert them into numpy arrays for each radionuclide.

So my view is the above is the best way of achieving what you want to do. Editing the code so inventories can accept either scalars or vectors in the dictionary when they are instantiated is not a good road to head down. Dragons lie down that way 🐉 🐲 - best to stay away!
